ast/src/analyzed/mod.rs

```diff
@@ -87,6 +89,109 @@ impl<T> Analyzed<T> {
+            to_remove: &BTreeSet<PolyID>,
+            shift: u64,
+            mut replacements: BTreeMap<PolyID, PolyID>,
+            poly: PolyID,
+        ) -> (u64, BTreeMap<PolyID, PolyID>) {
+            if to_remove.contains(&poly) {
+                (shift + 1, replacements)
+            } else {
+                replacements.insert(
+                    poly,
+                    PolyID {
+                        id: poly.id - shift,
+                        ..poly
+                    },
+                );
+                (shift, replacements)
+            }
+        }
+
+        // TODO intermediate polys
+        let witness_replacements = self
+            .committed_polys_in_source_order()
+            .iter()
+            .fold(
+                (0, BTreeMap::new()),
+                |(shift, replacements), (poly, _def)| {
+                    fold_replacement(to_remove, shift, replacements, poly.into())
+                },
+            )
+            .1;
+        let fixed_replacements = self
+            .constant_polys_in_source_order()
+            .iter()
+            .fold(
+                (0, BTreeMap::new()),
+                |(shift, replacements), (poly, _def)| {
+                    fold_replacement(to_remove, shift, replacements, poly.into())
+                },
+            )
+            .1;
+        let mut names_to_remove: HashSet<String> = Default::default();
+        self.definitions.retain(|name, (poly, _def)| {
+            if to_remove.contains(&(poly as &Polynomial).into()) {
+                names_to_remove.insert(name.clone());
+                false
+            } else {
+                true
+            }
+        });
+        self.source_order.retain(|s| {
+            if let StatementIdentifier::Definition(name) = s {
+                if names_to_remove.contains(name) {
+                    return true;
+                }
+            }
+            false
+        });
+        previsit_expressions_in_pil_file_mut(self, &mut |expr| {
+            if let Expression::PolynomialReference(poly) = expr {
+                let poly_id = poly.poly_id.unwrap();
+                assert!(!to_remove.contains(&poly_id));
+                match poly_id.ptype {
+                    PolynomialType::Committed => {
+                        poly.poly_id = Some(witness_replacements[&poly_id])
+                    }
+                    PolynomialType::Constant => poly.poly_id = Some(fixed_replacements[&poly_id]),
+                    PolynomialType::Intermediate => unimplemented!("Not yet implemented"),
+                }
+            }
+            ControlFlow::Continue::<()>(())
+        });
+    }
+
+    /// Adds a polynomial identity and returns the ID.
+    pub fn append_polynomial_identity(
+        &mut self,
+        identity: Expression<T>,
+        source: SourceRef,
+    ) -> u64 {
+        let id = self.identities.len() as u64;
+        self.identities.push(Identity {
+            id,
```

Comment:
Why identity uses `u64` here, and uses `usize` inside `StatementIdentifier::Identity`?

Reply:
Yeah, I don't know. In some places one is more convenient than the other...